File: backend/agent.py
from datetime import datetime
import generate
import evaldb
from template_router import load_templates, match_top_templates, run_matched_template
from intent import extract_intent, is_data_query, resolve_defaults, build_preamble, build_suggestions, build_intent_prompt_block
from verify import verify_rows


def make_timestamp_id():
	now = datetime.now()
	return now.strftime("%Y-%m-%d %H:%M:%S") + f".{now.microsecond * 1000:09d}"


# mirror every SSE event into a single `content` dict — same shape the frontend
# assembles live in App.jsx handleSubmit. this dict is persisted at stream end
# so that reloading chat history renders identically to live streaming.
def _collect(event, content):
	t = event.get("type")
	if t == "msg_id":
		content["msg_id"] = event["id"]
	elif t == "preamble":
		content["preamble"] = event["text"]
	elif t == "intent":
		content["intent"] = event["intent"]
	elif t == "token":
		content["streaming_text"] = (content.get("streaming_text") or "") + event["text"]
	elif t == "text":
		content["text"] = event["text"]
		content["raw_text"] = event["text"]
	elif t == "sql":
		content["sql"] = event["sql"]
		if event.get("explanation"):
			content["explanation"] = event["explanation"]
		if event.get("plot_config") is not None:
			content["plot_config"] = event["plot_config"]
		if content.get("streaming_text"):
			content["raw_text"] = content["streaming_text"]
		if content["rounds"]:
			content["rounds"][-1]["sql"] = event["sql"]
	elif t == "rows":
		content["columns"] = event["columns"]
		content["rows"] = event["rows"]
		if content["rounds"]:
			content["rounds"][-1]["columns"] = event["columns"]
			content["rounds"][-1]["rows"] = event["rows"]
	elif t == "explanation":
		content["explanation"] = event["text"]
	elif t == "summary":
		content["summary"] = event["text"]
	elif t == "suggestions":
		content["suggestions"] = event["items"]
	elif t == "key_takeaways":
		content["key_takeaways"] = event["items"]
	elif t == "plot_config":
		content["plot_config"] = event["plot_config"]
	elif t == "no_plot":
		content["no_plot"] = True
	elif t == "template_plots":
		content["template_plots"] = event["plots"]
	elif t == "distilled_summary":
		content["distilled_summary"] = event["text"]
	elif t == "round":
		content["rounds"].append({"label": event["label"]})
	elif t == "prompt":
		if content["rounds"]:
			content["rounds"][-1]["prompt"] = event["text"]
	elif t == "messages":
		if content["rounds"]:
			content["rounds"][-1]["messages"] = event["messages"]
	elif t == "response":
		if content["rounds"]:
			content["rounds"][-1]["response"] = event["text"]
	elif t == "tool_call":
		if content["rounds"]:
			r = content["rounds"][-1]
			if "tool_calls" not in r:
				r["tool_calls"] = []
			r["tool_calls"].append({"name": event["name"], "input": event["input"], "id": event["id"]})
	elif t == "tool_result":
		if content["rounds"]:
			for tc in content["rounds"][-1].get("tool_calls", []):
				if tc["id"] == event["id"]:
					tc["rows"] = event["rows"]
					break
	elif t == "user_prompt":
		content["user_prompt"] = event["text"]
	elif t == "error":
		content["error"] = event["error"]


# strong signals that indicate a self-contained new question
_STRONG_SIGNAL_KEYS = ("kpi_type", "service_ids", "top_n", "countries", "category", "service_filter")


def _last_assistant_ctx(history):
	# walk history in reverse, return the most recent assistant turn that has sql+intent attached
	for h in reversed(history or []):
		if h.get("role") == "assistant" and h.get("sql") and h.get("intent"):
			return h
	return None


def is_continuation(partial, prior_ctx):
	if prior_ctx is None:
		return False
	if not is_data_query(partial):
		return True
	has_strong = any(partial.get(k) for k in _STRONG_SIGNAL_KEYS)
	if has_strong:
		return False
	return True


def _merge_partial_over_intent(prior_intent, partial):
	merged = {}
	for k in ("kpi_type", "kpi_dimension", "kpi_detail", "category", "countries",
			  "service_ids", "service_filter", "top_n", "year", "quarter",
			  "trend_mode", "age_group", "population_segment", "service_level",
			  "video_type_comparison"):
		if prior_intent.get(k) is not None:
			merged[k] = prior_intent[k]
	for k, v in partial.items():
		merged[k] = v
	return merged


async def generate_agent_stream(prompt, history=None, user="", conversation_id=""):
	content = {"loading": False, "rounds": []}
	try:
		async for event in _generate_agent_stream_inner(prompt, history, user, conversation_id):
			_collect(event, content)
			yield event
	finally:
		msg_id = content.get("msg_id")
		if msg_id:
			try:
				evaldb.update_result_data(msg_id, content)
			except Exception as e:
				print(f"[agent] persist content failed: {e}")


async def _generate_agent_stream_inner(prompt, history=None, user="", conversation_id=""):
	if history is None:
		history = []

	if not conversation_id:
		conversation_id = make_timestamp_id()
	yield {"type": "conversation_id", "id": conversation_id}

	msg_id = make_timestamp_id()
	yield {"type": "msg_id", "id": msg_id}
	yield {"type": "user_prompt", "text": prompt}

	# stage 0: intent resolution — extract + resolve defaults
	partial = extract_intent(prompt)

	# continuation detection — default to continue unless clearly a new question
	prior_ctx = _last_assistant_ctx(history)
	continuation = is_continuation(partial, prior_ctx)
	prior_sql = None
	prior_plot_config = None
	if continuation:
		prior_sql = prior_ctx.get("sql")
		prior_plot_config = prior_ctx.get("plot_config")
		prior_intent = prior_ctx.get("intent") or {}
		partial = _merge_partial_over_intent(prior_intent, partial)
		print(f"[agent] continuation detected — merged partial: {partial}")

	intent = None
	intent_block = ""
	if is_data_query(partial):
		intent = resolve_defaults(partial)
		preamble = build_preamble(intent)
		intent_block = build_intent_prompt_block(intent)
		print(f"[intent] resolved: kpi={intent['kpi_type']} dim={intent.get('kpi_dimension')} countries={intent.get('countries')} defaults={intent.get('applied_defaults')}")
		yield {"type": "preamble", "text": preamble}
		yield {"type": "intent", "intent": intent}

	# on continuation, skip template routing — go straight to tool-loop so LLM can modify prior SQL
	templates = {}
	matches = []
	match_debug = None
	template_fallback_feedback = None
	if not continuation:
		templates = load_templates()
		if templates:
			matches, match_debug = await match_top_templates(prompt, templates)

		# stage 1: routing — always surface the Haiku call, even when it returned NONE / errored
		if match_debug:
			yield {"type": "round", "label": "Routing"}
			yield {"type": "prompt", "text": match_debug["prompt"]}
			yield {"type": "messages", "messages": match_debug["messages"]}
			yield {"type": "response", "text": match_debug["response"] or "(no response)"}
		if matches and matches[0]["score"] >= 0.95:
			yield {"type": "round", "label": "Template Execution"}
			distilled = ""
			template_events = []
			template_cols = None
			template_rows = None
			async for event in run_matched_template({
				"prompt": prompt,
				"match": matches[0],
				"template": templates[matches[0]["file"]],
				"msg_id": msg_id,
				"user": user,
				"conversation_id": conversation_id,
				"intent": intent,
			}):
				if event.get("type") == "summary":
					distilled = event.get("text", "")
				elif event.get("type") == "text" and not distilled:
					distilled = event.get("text", "")[:500]
				elif event.get("type") == "rows":
					template_cols = event["columns"]
					template_rows = event["rows"]
				elif event.get("type") == "error":
					template_rows = []
				template_events.append(event)

			verdict = await verify_rows(prompt, template_cols, template_rows or [])
			print(f"[agent] template verify: ok={verdict['ok']} reason={verdict['reason']}")
			if verdict["ok"]:
				for e in template_events:
					yield e
				if intent:
					yield {"type": "suggestions", "items": build_suggestions(intent)}
				if distilled:
					yield {"type": "distilled_summary", "text": distilled}
				return
			template_fallback_feedback = verdict["reason"]

	# stage 2: tool-loop generation — LLM with `query` tool, optional template hints
	if continuation:
		label = "Follow-up Generation"
	elif matches:
		label = "Guided Generation"
	else:
		label = "Open Generation"
	distilled = ""
	suggestions_yielded = False
	async for event in generate.run({
		"prompt": prompt,
		"matches": matches,
		"templates": templates,
		"history": history,
		"msg_id": msg_id,
		"user": user,
		"conversation_id": conversation_id,
		"intent_block": intent_block,
		"prior_sql": prior_sql,
		"prior_plot_config": prior_plot_config,
		"label": label,
		"template_fallback_feedback": template_fallback_feedback,
	}):
		if event.get("type") == "summary":
			distilled = event.get("text", "")
		elif event.get("type") == "text" and not distilled:
			distilled = event.get("text", "")[:500]
		elif event.get("type") == "suggestions":
			suggestions_yielded = True
		yield event
	if intent and not suggestions_yielded:
		yield {"type": "suggestions", "items": build_suggestions(intent)}
	if distilled:
		yield {"type": "distilled_summary", "text": distilled}

File: node/agent.js
import { updateResultData } from './sqlite.js'
import { loadTemplates, matchTopTemplates, runMatchedTemplate } from './template_router.js'
import { buildSystemPrompt, openGeneration, getServiceIdToCanonical } from './generate2.js'
import {validateColumnCardinality} from "./Cardinality.js"

import { saveTemplateFromContent } from './template_save.js'
import { metricProbe, MAX_ROWS } from './metricProbe.js'
import { summaryReport } from './summaryReport.js'
import { clarificationReport, notAvailableReport } from './fallbackReport.js'
import { executeQuery } from './db.js'
import { buildMessages } from './sql_utils.js'

const MAX_ATTEMPTS = 4
function _makeTimestampId() {
	const now = new Date()
	const ts = now.toISOString().replace('T', ' ').slice(0, 19)
	const ns = (now.getMilliseconds() * 1000000).toString().padStart(9, '0')
	return `${ts}.${ns}`
}

// mirrors every SSE event into a single content dict — same shape the frontend
// assembles live in App.jsx handleSubmit. persisted at stream end for history replay.
export function _collect(event, content) {
	const t = event.type
	if (t === 'msgId') {
		content.msgId = event.id
	} else if (t === 'preamble') {
		content.preamble = event.text
	} else if (t === 'intent') {
		content.intent = event.intent
	} else if (t === 'token') {
		content.streamingText = (content.streamingText || '') + event.text
	} else if (t === 'text') {
		content.text = event.text
		content.rawText = event.text
	} else if (t === 'sql') {
		content.sql = event.sql
		if (event.explanation) content.explanation = event.explanation
		if (event.plotConfig !== undefined && event.plotConfig !== null) content.plotConfig = event.plotConfig
		if (content.streamingText) content.rawText = content.streamingText
		if (content.rounds.length > 0) content.rounds[content.rounds.length - 1].sql = event.sql
	} else if (t === 'rows') {
		content.columns = event.columns
		content.rows = event.rows
		if (content.rounds.length > 0) {
			content.rounds[content.rounds.length - 1].columns = event.columns
			content.rounds[content.rounds.length - 1].rows = event.rows
		}
	} else if (t === 'explanation') {
		content.explanation = event.text
	} else if (t === 'summary') {
		content.summary = event.text
	} else if (t === 'suggestions') {
		content.suggestions = event.items
	} else if (t === 'keyTakeaways') {
		content.keyTakeaways = event.items
	} else if (t === 'plotConfig') {
		content.plotConfig = event.plotConfig
	} else if (t === 'noPlot') {
		content.noPlot = true
	} else if (t === 'answerType') {
		content.answerType = event.answerType
	} else if (t === 'clarification') {
		content.clarification = event.text
	} else if (t === 'cards') {
		content.cards = event.items
	} else if (t === 'templatePlots') {
		content.templatePlots = event.plots
	} else if (t === 'distilledSummary') {
		content.distilledSummary = event.text
	} else if (t === 'round') {
		content.rounds.push({ label: event.label })
	} else if (t === 'prompt') {
		if (content.rounds.length > 0) {
			content.rounds[content.rounds.length - 1].prompt = event.text
		}
	} else if (t === 'messages') {
		if (content.rounds.length > 0) {
			content.rounds[content.rounds.length - 1].messages = event.messages
		}
	} else if (t === 'response') {
		if (content.rounds.length > 0) {
			content.rounds[content.rounds.length - 1].response = event.text
		}
	} else if (t === 'tool_call') {
		if (content.rounds.length > 0) {
			const r = content.rounds[content.rounds.length - 1]
			if (!r.tool_calls) r.tool_calls = []
			r.tool_calls.push({ name: event.name, input: event.input, id: event.id })
		}
	} else if (t === 'tool_result') {
		if (content.rounds.length > 0) {
			const tc = (content.rounds[content.rounds.length - 1].tool_calls || []).find(c => c.id === event.id)
			if (tc) tc.rows = event.rows
		}
	} else if (t === 'userPrompt') {
		content.userPrompt = event.text
	} else if (t === 'error') {
		content.error = event.error
	}
}

// history is send from client side
export async function* generateAgentStream(prompt, history, options) {
	const { user = '', conversationId = '' } = options || {}
	const content = { loading: false, rounds: [] }
	try {
		for await (const event of _generateAgentStreamInner(prompt, history, user, conversationId)) {
			_collect(event, content)
			yield event
		}
	} finally {
		const msgId = content.msgId
		if (msgId) {
			try {
				updateResultData(msgId, content)
			} catch (e) {
				console.log('[agent] persist content failed:', e.message)
			}
			if (content.sql && (content.plotConfig || content.templatePlots) && !content.error) {
				try {
					const path = saveTemplateFromContent(content)
					if (path) console.log('[agent] auto-saved template', path)
				} catch (e) {
					console.log('[agent] auto-save template failed:', e.message)
				}
			}
		}
	}
}

async function* _generateAgentStreamInner(userPrompt, history, user, conversationId) {
	const hist = history || []
	conversationId = conversationId || _makeTimestampId()
	console.log(`[agent] start user=${user} conversationId=${conversationId} histLen=${hist.length}`)
	yield { type: 'conversationId', id: conversationId }

	const msgId = _makeTimestampId()
	yield { type: 'msgId', id: msgId }
	yield { type: 'userPrompt', text: userPrompt }

	let priorSql = null
	let priorPlotConfig = null
	for (let i = hist.length - 1; i >= 0; i--) {
		const h = hist[i]
		if (h.role === 'assistant' && h.sql) {
			priorSql = h.sql
			priorPlotConfig = h.plotConfig || null
			break
		}
	}

	const templates = loadTemplates()
	const hasTemplates = Object.keys(templates).length > 0

	const [probeResult, matchResult] = await Promise.all([
		metricProbe(userPrompt),
		hasTemplates ? matchTopTemplates(userPrompt, templates) : Promise.resolve([[], null]),
	])
	console.log(`[agent] probe: answer_type=${probeResult.answer_type} answer_confidence=${probeResult.answer_confidence} candidates=${probeResult.candidates.length}`)

	if (probeResult.answer_type === 'clarification_needed') {
		for await (const e of clarificationReport(userPrompt)) yield e
		return
	}
	if (probeResult.answer_type === 'data_not_available') {
		for await (const e of notAvailableReport(userPrompt)) yield e
		return
	}

	const _PROBE_TO_DISPLAY = {
		ranking: 'table', comparison: 'table', distribution: 'table',
		trend: 'trend',
		text: 'text', correlation: 'text', market_overview: 'text',
	}
	const answerType = _PROBE_TO_DISPLAY[probeResult.answer_type] || 'text'
	yield { type: 'answerType', answerType }

	let matches = []
	if (matchResult[1]) {
		matches = matchResult[0]
		const matchDebug = matchResult[1]
		yield { type: 'round', label: 'Routing' }
		yield { type: 'prompt', text: matchDebug.prompt }
		yield { type: 'messages', messages: matchDebug.messages }
		yield { type: 'response', text: matchDebug.response || '(no response)' }
	}

	// fast path: template with high confidence score
	if (matches.length > 0 && matches[0].score >= 0.95) {
		yield { type: 'round', label: 'Template Execution' }
		let templateCols = null
		let templateRows = null
		let templateSql = null
		let gotError = false

		for await (const event of runMatchedTemplate({
			userPrompt,
			match: matches[0],
			template: templates[matches[0].file],
			msgId,
			user,
			conversationId,
			history: hist,
			probe: probeResult,
		})) {
			if (event.type === 'rows') { templateCols = event.columns; templateRows = event.rows }
			else if (event.type === 'sql') { templateSql = event.sql }
			else if (event.type === 'error') { gotError = true }
			yield event
		}

		if (!gotError && templateRows && templateRows.length > 0) {
			let summaryOk = false
			for await (const e of summaryReport({
				userPrompt, columns: templateCols, rows: templateRows, sql: templateSql,
				answerType, priorPlotConfig, msgId, user, conversationId,
			})) {
				if (e.type === '_summary_status') { summaryOk = e.ok; continue }
				yield e
			}
			if (summaryOk) return
		}
	}

	// median path: try all high-confidence probe candidates in confidence order
	const bestCandidates = probeResult.candidates
		.filter(c => c.answer_confidence >= 0.8 && c.row_count > 0 && c.row_count < MAX_ROWS)
		.sort((a, b) => b.answer_confidence - a.answer_confidence)

	const probeCols = ['period_date', 'country', 'kpi_type', 'kpi_dimension', 'service_id', 'age_group', 'value']
	for (let i = 0; i < bestCandidates.length; i++) {
		const candidate = bestCandidates[i]
		console.log(`[agent] median path [${i + 1}/${bestCandidates.length}]: answer_confidence=${candidate.answer_confidence} kpi_type=${candidate.kpi_type}`)
		yield { type: 'round', label: 'Probe Answer' }
		if (i === 0 && probeResult.llm_prompt) yield { type: 'prompt', text: probeResult.llm_prompt }
		if (i === 0 && probeResult.llm_response) yield { type: 'response', text: probeResult.llm_response }
		yield { type: 'sql', sql: candidate.sql }
		yield { type: 'rows', columns: probeCols, rows: candidate.rows }

		let summaryOk = false
		for await (const e of summaryReport({
			userPrompt, columns: probeCols, rows: candidate.rows, sql: candidate.sql,
			answerType, priorPlotConfig, msgId, user, conversationId,
		})) {
			if (e.type === '_summary_status') { summaryOk = e.ok; continue }
			yield e
		}
		if (summaryOk) return
	}

	// slow path: retry loop with SQL generation
	const slowLabel = matches.length > 0 ? 'Guided Generation' : 'Open Generation'
	console.log(`[agent] slow path label=${slowLabel} matches=${matches.length}`)

	const systemPrompt = await buildSystemPrompt({ matches, templates, priorSql, probe: probeResult })
	const messages = buildMessages(hist, userPrompt)
	let lastSql = null, lastColumns = null, lastRows = null, lastFailReason = null

	for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
		const roundLabel = attempt === 0 ? slowLabel : `Retry ${attempt}`
		let genResult = null
		for await (const e of openGeneration({
			system: systemPrompt, messages, label: roundLabel,
			msgId, user, conversationId,
		})) {
			if (e.type === '_gen_result') { genResult = e; continue }
			yield e
		}

		if (genResult.kind === 'text') return

		const { sql, fullText } = genResult
		let columns, rows
		try {
			const result = await executeQuery(sql)
			columns = result.columns
			rows = result.rows

			yield { type: 'sql', sql, plotConfig: null, explanation: '' }
			yield { type: 'rows', columns, rows }

			if (rows.length === 0) {
				lastFailReason = 'Query returned no rows'
				console.log('[agent] query returned 0 rows')
				yield { type: 'error', error: 'Query returned no rows' }
				messages.push(
					{ role: 'assistant', content: fullText },
					{ role: 'user', content: `## Revision Feedback\nThe query returned 0 rows. The data may not exist for these filter criteria. Try different filters, a different time period, or a different metric.\n\nWrite a new SQL query.` }
				)
				continue
			}

			const cardinalityCheck = validateColumnCardinality(columns, rows)
			rows = rows.map(r => Object.fromEntries(columns.map(col => [col, r[col]])))

			if (cardinalityCheck.ok) {
				if (columns.includes('service_id')) {
					const map = await getServiceIdToCanonical()
					columns = [...columns, 'canonical_name']
					rows = rows.map(r => ({ ...r, canonical_name: map[r.service_id] || null }))
				}
			} else {
				lastFailReason = cardinalityCheck.reason
				console.log('[agent] cardinality check failed:', cardinalityCheck.reason)
				yield { type: 'error', error: cardinalityCheck.reason }
				messages.push(
					{ role: 'assistant', content: fullText },
					{ role: 'user', content: `## Revision Feedback\n${cardinalityCheck.reason}\n\nRewrite the SQL to fix this.` }
				)
				continue
			}
		} catch (e) {
			console.log('[agent] query error:', e.message)
			lastFailReason = e.message
			yield { type: 'error', error: `SQL error: ${e.message}` }
			messages.push(
				{ role: 'assistant', content: fullText },
				{ role: 'user', content: `## Revision Feedback\nSQL error: ${e.message}\n\nFix the SQL and try again.` }
			)
			continue
		}

		const sqlHistory = [...messages, { role: 'assistant', content: fullText }]
		let summaryOk = false
		for await (const e of summaryReport({
			userPrompt, columns, rows, sql,
			answerType, priorPlotConfig, msgId, user, conversationId,
			sqlHistory,
		})) {
			if (e.type === '_summary_status') { summaryOk = e.ok; continue }
			yield e
		}
		if (summaryOk) return

		console.log(`[agent] attempt ${attempt + 1} verify failed`)
		lastSql = sql; lastColumns = columns; lastRows = rows
		lastFailReason = "Data doesn't answer the question"
		messages.push(
			{ role: 'assistant', content: fullText },
			{ role: 'user', content: `## Revision Feedback\nThe data does not answer the question.\n\nWrite a new SQL query.` }
		)
	}

	// best effort after exhausting retries
	if (lastColumns && lastRows) {
		for await (const e of summaryReport({
			userPrompt, columns: lastColumns, rows: lastRows, sql: lastSql,
			answerType, priorPlotConfig, msgId, user, conversationId,
			force: true,
		})) {
			if (e.type === '_summary_status') continue
			yield e
		}
	} else {
		const text = lastFailReason
			? `I was unable to find data that answers your question. ${lastFailReason}`
			: `I was unable to find data that answers your question after several attempts.`
		yield { type: 'summary', text }
	}

	console.log(`[agent] _generateAgentStreamInner done`)
}
